docs/design/2021-03-01-pipelined-window-functions.md

```diff
@@ -0,0 +1,111 @@
+
+### Pipelining
+
+* assume the total number of rows in a partition is N, which we do not know in advance since the data is pipelined  
+  * UpdatePartialResult: append partial rows to the aggregation function, needs to append N rows eventually
+  * Slide (perhaps this function is better implemented on windowProcessor): must be called before calling AppendFinalResult2Chunk, it returns success or fail if the current rows consumed by UpdatePartialResult is not enough to slide for next row, it will also return an estimated number of rows so that it can be called (useful for unbounded following, we can use -1 to denote that it needs the whole partition, 0 means success immediately, and n if we could know the number of rows needed (for rowFrame) or 1 for rangeFrame since we need to examine row by row
+  * AppendFinalResult2Chunk: append result for one row, can be called N times, and can only be called after a successful slide
+  * FinishUpdate: called upon the whole partition has been appended, this is to notify the SlidingWindowAggFunc that the whole partition is consumed, so that for those function that needs the whole partition, it is now time to return success on slide
+* We want the movement of the sliding window to be the driver for data fetching on the children executor, so the dataflow logic needs to be modified  
+  * Next() will call windowProcessor’s Slide function to drive it
+  * Slide function will fetch data from child, and use vecGroupCheck to split it
+  * Then the data is processed at maximum effort using  
+    * UpdatePartialResult
+    * Slide or do nothing if it is not SlidingWindowAggFunc
+    * If Slide returns success or the whole partition is processed  
+      * Obtain the result using AppendFinalResult2Chunk
+  * Result is then feed back to Next, and returned once the chunk is full  
+    * There could be a background goroutine pulling data
+
+## Rationale
+
+This feature will decrease memory consumption for executing window function.
+
+## Compatibility
+
+Pipelining won't cause any compatibility issue.
+
+## Implementation
+
+* [x] Create PipelinedWindowExec based on current implementation and modify the windowProcessor interface (see [PR23022](https://github.com/pingcap/tidb/pull/23022)).
+* [x] Change data flow, make Next() pulling data from windowProcessor, and windowProcessor calls fetchChild and process data at maximum effort.
+* [x] Modify Slide semantic and add FinishUpdate function on SlidingWindowAggFunc interface, and modify correspondingly on each window function.
+* [x] Done pipelining for SlidingWindowAggFunc, add test to make sure it is correct.
+* [x] Modify RN to be SlidingWindowAggFunc, and add planner support.
+* [x] Add test for RN.
+* [x] Benchmark, make sure it has constant memory consumption and no execution time regression.
```

Comment:
@ichn-hu I'd like to know where these rest items implemented.

Reply:
they are implemented in a single pr

Reply:
Thanks! Does this commit 4755af84a10ab790297a81a787afb7b9ec8300c3 speak out the truth?

Reply:
sure